contracts/TokenVesting.sol

```diff
@@ -2,355 +2,442 @@
+        bytes32 vestingScheduleId = computeNextVestingScheduleIdForHolder(
+            _beneficiary
+        );
+        unchecked {
+            uint32 cliff = _start + _cliff;
+            require(cliff >= _start);
+            require(_start + _duration > _start);
+            vestingSchedules[vestingScheduleId] = VestingSchedule({
+                initialized: true,
+                revocable: _revocable,
+                revoked: false,
+                beneficiary: _beneficiary,
+                cliff: cliff,
+                start: _start,
+                duration: _duration,
+                amountTotal: _amount,
+                released: 0
+            });
+            // Cannot realistically overflow
+            uint256 _vestingSchedulesTotalAmount = vestingSchedulesTotalAmount;
+            uint256 totalAmount = _vestingSchedulesTotalAmount + _amount;
+            require(totalAmount > _vestingSchedulesTotalAmount);
+            vestingSchedulesTotalAmount = totalAmount;
+            vestingSchedulesIds.push(vestingScheduleId);
+            holdersVestingCount[_beneficiary] += 1;
+        }
+        emit Created(vestingScheduleId, _beneficiary, _amount);
     }
 
     /**
-    * @dev Returns the last vesting schedule for a given holder address.
-    */
-    function getLastVestingScheduleForHolder(address holder)
-        public
-        view
-        returns(VestingSchedule memory){
-        return vestingSchedules[computeVestingScheduleIdForAddressAndIndex(holder, holdersVestingCount[holder] - 1)];
+     * @notice Revokes the vesting schedule for given identifier.
+     * @param vestingScheduleId the vesting schedule identifier
+     */
+    function revoke(
+        bytes32 vestingScheduleId
+    ) external onlyOwner onlyIfVestingScheduleNotRevoked(vestingScheduleId) {
+        VestingSchedule storage vestingSchedule = vestingSchedules[
+            vestingScheduleId
+        ];
+        if (!vestingSchedule.revocable) revert NotRevocable();
+        uint112 vestedAmount = _computeReleasableAmount(vestingSchedule);
+        if (vestedAmount != 0) {
+            release(vestingScheduleId, vestedAmount);
+        }
+        unchecked {
+            uint112 remainingAmount = vestingSchedule.amountTotal -
+                vestingSchedule.released;
+            vestingSchedulesTotalAmount -= remainingAmount;
+            vestingSchedule.revoked = true;
+            emit Revoked(vestingScheduleId, remainingAmount);
+        }
     }
 
     /**
-    * @dev Computes the vesting schedule identifier for an address and an index.
-    */
-    function computeVestingScheduleIdForAddressAndIndex(address holder, uint256 index)
-        public
-        pure
-        returns(bytes32){
-        return keccak256(abi.encodePacked(holder, index));
+     * @notice Withdraw the specified amount if possible.
+     * @param amount the amount to withdraw
+     */
+    function withdraw(uint256 amount) external onlyOwner {
```

Comment:
Why do we have a function to allow the owner of this contract to withdraw tokens? Shouldn't it be the case for the beneficiary to withdraw their own tokens?

Reply:
It is for the owner to withdraw excessive tokens not needed for vesting.